File: src/test/java/pl/tsimur/jpatest/repository/OrganizationRepositoryTest.java
package pl.tsimur.jpatest.repository;


import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.transaction.annotation.Transactional;
import pl.timur.jpatest.model.Organization;
import pl.timur.jpatest.model.User;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
public class OrganizationRepositoryTest {
    @Autowired
    private OrganizationRepository repositoryOrg;

    @Test
    void createTest() {
        // 1 - C: Create
        Organization organization = repositoryOrg.save(TestData.org1);

        // 2 - R: Read
        assertTrue(
                repositoryOrg.findById(TestData.org1.getNip()) // Возвращает Optional<Organization> - поэтому нужно распаковать .get()
                        .isPresent());
        repositoryOrg.deleteById(TestData.org1.getNip());
    }

    @Test
    @Transactional
    void CRUDTest() {
        // 1 - C: Create
        Organization organization = repositoryOrg.save(TestData.org1);
        System.out.println("🤖 Сохранил организацию: " + organization.getName());
        Integer id = organization.getNip();
        assertNotNull(id);

        // 2 - R: Read
        System.out.println("😺 Проверяю что организация появилась в Базе Данных...");
        assertNotNull(
                repositoryOrg.findOrganizationByNip(id) // Возвращает Optional<Organization> - поэтому нужно распаковать .get()
        );
//
        System.out.println("😹 Обновляем организации поле Name...");
        repositoryOrg.updateOrganizationByNip(organization.getNip(), "Stars");
//            assertEquals("Stars", organization.getName());

        // 4 - D: Delete
        System.out.println("😹 Удаляю организацию из базе данных...");
        repositoryOrg.deleteByNip(id);

        // И еще раз перепроверю
        System.out.println("😼 Проверяю что организации не должно быть после удаления из базе данных...");
        assertNull(
                repositoryOrg.findOrganizationByNip(id)
        );
    }

    @Test
    void findOrganizationByNameTest() {
        var orgName = "Mercedes";
        List<Organization> organizationList = repositoryOrg.findAllOrganizationByName(orgName);
        assertEquals(orgName, organizationList.get(0).getName());
    }

    @Test
    void findOrganizationByAddressTest() {
        String address = "ul.Pobeda, Munich";
        List<Organization> organizationList = repositoryOrg.findAllOrganizationByAddress(address);
        assertEquals(address, organizationList.get(0).getAddress());
    }

    @Test
    @Disabled
    void updateOrganizationByOwnerIdTest() {
        Integer nip = TestData.org1.getNip();
        String newName = "Oracle";

        Organization organization = repositoryOrg.save(TestData.org1);

        String oldName = organization.getName();

        assertEquals(oldName, organization.getName());

        repositoryOrg.updateOrganizationByNip(nip, newName);
        Organization organizationWithUpdatedName = repositoryOrg.findById(nip)
                .orElseThrow(() -> new RuntimeException("Not found organization with nip: " + nip));

        assertEquals(newName, organizationWithUpdatedName.getName());
        repositoryOrg.deleteByNip(nip);
    }

    static class TestData {
        static final User user1 = User.builder().id(-1).build();

        static final Organization org1 = Organization.builder()
                .nip(5674523)
                .address("ul.Slusara, Poznan")
                .name("Pull&Bear")
                .owner(user1)
                .build();
    }
}

File: src/test/kotlin/pl/timur/jpatest/repository/OrganizationRepositoryTest.kt
package pl.timur.jpatest.repository

import org.junit.jupiter.api.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.transaction.annotation.Transactional
import pl.timur.jpatest.model.Organization
import pl.timur.jpatest.model.User

@DataJpaTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class OrganizationRepositoryTest(@Autowired private val repositoryOrg: OrganizationRepository) {

    companion object {
        private val user1 = User(id = -1)

        val org1 = Organization(
            nip = 5674523,
            address = "ul.Slusara, Poznan",
            name = "Pull&Bear"
        )
    }


    @BeforeAll
    fun init() {
        org1.owner = user1
    }

    @Test
    fun createTest() {
        // 1 - C: Create
        val organization = repositoryOrg.save(org1)

        // 2 - R: Read
        Assertions.assertTrue(
            repositoryOrg.findById(org1.nip) // Возвращает Optional<Organization> - поэтому нужно распаковать .get()
                .isPresent
        )
        repositoryOrg.deleteById(org1.nip)
    }

    @Test
    @Transactional
    fun CRUDTest() {
        // 1 - C: Create
        val organization = repositoryOrg.save(org1)
        println("🤖 Сохранил организацию: " + organization.name)
        val id = organization.nip
        Assertions.assertNotNull(id)

        // 2 - R: Read
        println("😺 Проверяю что организация появилась в Базе Данных...")
        Assertions.assertNotNull(
            repositoryOrg.findOrganizationByNip(id) // Возвращает Optional<Organization> - поэтому нужно распаковать .get()
        )
        //
        println("😹 Обновляем организации поле Name...")
        repositoryOrg.updateOrganizationByNip(organization.nip, "Stars")
        //            assertEquals("Stars", organization.getName());

        // 4 - D: Delete
        println("😹 Удаляю организацию из базе данных...")
        repositoryOrg.deleteByNip(id)

        // И еще раз перепроверю
        println("😼 Проверяю что организации не должно быть после удаления из базе данных...")
        Assertions.assertNull(
            repositoryOrg.findOrganizationByNip(id)
        )
    }

    @Test
    fun findOrganizationByNameTest() {
        val orgName = "Mercedes"
        val organizationList = repositoryOrg.findAllOrganizationByName(orgName)
        Assertions.assertEquals(orgName, organizationList[0].name)
    }

    @Test
    fun findOrganizationByAddressTest() {
        val address = "ul.Pobeda, Munich"
        val organizationList = repositoryOrg.findAllOrganizationByAddress(address)
        Assertions.assertEquals(address, organizationList[0].address)
    }

    @Test
    @Disabled
    fun updateOrganizationByOwnerIdTest() {
        val nip = org1.nip
        val newName = "Oracle"
        val organization = repositoryOrg.save(org1)
        val oldName = organization.name
        Assertions.assertEquals(oldName, organization.name)
        repositoryOrg.updateOrganizationByNip(nip, newName)
        val organizationWithUpdatedName = repositoryOrg.findById(nip)
            .orElseThrow { RuntimeException("Not found organization with nip: $nip") }
        Assertions.assertEquals(newName, organizationWithUpdatedName.name)
        repositoryOrg.deleteByNip(nip)
    }

}